### PR TITLE
ga4gh_benchmark server benchmarking

### PR DESCRIPTION
**Pull request by Naburimannu**
Here's a more substantive strawman for #46.

Open issues:
* How many queries?
* Which queries?

### CODE REVIEW

**Comment by Naburimannu** [comment]
Just to be explicit: if you have the ga4gh-example-data (Wormtable) dataset installed, you can type 'ga4gh_benchmark' and get a number ~15 minutes later.

Running this verified for me that the memory bloat issue (#10) is still open.

**Comment by kerrydc** [comment]
+1

File: ga4gh/cli.py [hunk 1]
```diff
@@ -63,6 +64,51 @@ def server_main():
...
+    initialRequest.start = 0
+    initialRequest.end = 100000
+    initialRequestString = initialRequest.toJSONString()
+
+    # Repeat the query ~10x; if we compute average we should probably
+    # throw out at least the first one.
+    # TODO: more sophisticated statistics. Sometimes we want min(),
+    # sometimes mean = sum() / len(), sometimes other measures,
+    # perhaps exclude outliers...
+    for i in range(0, 10):
```
Comment (jeromekelleher) [inline ga4gh/cli.py]: The `0` isn't necessary here.

File: ga4gh/cli.py [hunk 2]
```diff
@@ -63,6 +64,51 @@ def server_main():
...
+        start_time = time.clock()
```
Comment (jeromekelleher) [inline ga4gh/cli.py]: camelCase please.

File: ga4gh/cli.py [hunk 3]
```diff
@@ -63,6 +64,51 @@ def server_main():
         app.backend = backend
         app.run(host="0.0.0.0", port=args.port, debug=True)
 
+
+def server_benchmark():
```
Comment (jeromekelleher) [inline ga4gh/cli.py]: This should probably be `server_benchmark_main()`, just to keep the convention that the main entry point for any program will end with `main`. (We should camelCase this convention at some point though)

**Comment by jeromekelleher** [comment]
I like it @Naburimannu, we definitely want this. I wonder though, do we want to ship this as a first-class CLI program installed on every user's machine, or should it be a development tool (say, `server_benchmark.py`)? If it is to be a first-class program, then the input will need to be generalised and we'll need to provide a UI to set the queries and so on. If it's a dev script, then I'm happy for it to depend on example data living in a fixed location. What do you think?

File: ga4gh/cli.py (hunk 1 above)
Comment (Naburimannu) [inline ga4gh/cli.py]: Agreed; when I was chasing down a bug I made everything explicit, but it doesn't need to be shipped.

**Comment by Naburimannu** [comment]
Coverted to a standalone script with a header comment to tell you (briefly) what data we want where. Started parameterizing it for future plans.

Before the b2bd475 commit, it would take ~90 seconds to run on my desktop to time ~30 seconds inside Backend.searchVariants(), because the JSON parsing overhead is huge. Using a regexp instead feels fragile, but 3x speedup for a tool makes this a lot more feasible to consider running regularly, and this is explicitly the "dev script" branch of of @jeromekelleher's question so we might tolerate some fragility.

File: server_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+def timeOneSearch(queryString):
+    """
+    Returns (search result as JSON string, time elapsed during search)
+    """
+    startTime = time.clock()
+    resultString = backend.searchVariants(queryString)
+    endTime = time.clock()
+    elapsedTime = endTime - startTime
+    return (resultString, elapsedTime)
```
Comment (jeromekelleher) [inline server_benchmark.py]: The brackets aren't actually needed here. Feel free to keep them if you think it improves readability --- just thought I'd point it out.

File: server_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+    # TODO: more sophisticated statistics. Sometimes we want min(),
+    # sometimes mean = sum() / len(), sometimes other measures,
+    # perhaps exclude outliers...
+
+    # If we compute average we should throw out at least the first one.
+    # return sum(times[2:])/len(times[2:])
+    return min(times)
+
+
```
Comment (jeromekelleher) [inline server_benchmark.py]: I'd rather see a `if __name__ == '__main__'` condition here.

**Comment by jeromekelleher** [comment]
Looks great @Naburimannu. Other than a couple of tiny (optional) nits, I think it's perfect. I agree on the use of the regex - JSON parsing is a massive overhead, and just distracts from what we're really trying to measure.

**Comment by Naburimannu** [comment]
Nits addressed and squashed; ready to review? (the CORS test was broken by a previous commit; filed https://github.com/ga4gh/server/issues/94 with my attempt at diagnosis)

**Comment by jeromekelleher** [comment]
+1, thanks @Naburimannu. @dcolligan, would you mind having a quick look over this and +1'ing if you're happy with it?

I'm looking forward to seeing memory profiling and the other great tools you've developed getting integrated into this in later PRs!

File: server_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    # perhaps exclude outliers...
+
+    # If we compute average we should throw out at least the first one.
+    # return sum(times[2:])/len(times[2:])
+    return min(times)
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
+        description="GA4GH reference server benchmark")
+    args = parser.parse_args()
```
Comment (dcolligan) [inline server_benchmark.py]: nit: eliminate unused args variable

File: server_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        accruedTime = elapsedTime
+        pageCount = 1
+        token = extractNextPageToken(resultString)
+        # Iterate to go beyond the first page of results.
+        while token is not None and pageCount < pageLimit:
+            pageRequest = request
+            pageRequest.pageToken = token
+            pageRequestString = pageRequest.toJSONString()
+            resultString, elapsedTime = timeOneSearch(pageRequestString)
+            accruedTime = accruedTime + elapsedTime
```
Comment (dcolligan) [inline server_benchmark.py]: nit: could do accruedTime += elapsedTime

File: ga4gh/cli.py [hunk 8]
```diff
@@ -66,6 +66,7 @@ def server_main():
             args.dataDir, args.variantSetClass)
         frontend.app.run(host="0.0.0.0", port=args.port, debug=True)
 
+
```
Comment (dcolligan) [inline ga4gh/cli.py]: why this change?  generally, let's try to keep the diff as small as possible, and especially not add unnecessary files to the diff (unless this is some pep8 thing?)

**Comment by dcolligan** [comment]
The only thing that is keeping me from +1'ing this is the file path of `server_benchmark.py` ... do we want it in the root dir?  I think we should keep a few things in the root dir as possible.  What about under scripts/ or perhaps creating a new dir dev_scripts/ ?

**Comment by Naburimannu** [comment]
I have no objection to moving it to scripts/; it's where it is to mimic server_dev.py and client_dev.py.
I'll clean up the nits, and agree that the blank line shouldn't be there; will remove unless it's a PEP8 issue.

**Comment by jeromekelleher** [comment]
It's a lot easier to keep these scripts in the root, as you'll have to install the full `ga4gh` package if you want to run them otherwise (they import ga4gh stuff). The only simple way I found to do it (without installing) was to create a hard link from `scripts/something.py` to `something.py` in the root, and then  run `python something.py`. However, these hard links kept breaking when git changed files around, so I was never quite sure what version of the file I was working with. Unless someone's got a better workaround for the import problem, I'd much rather leave scripts that import `ga4gh` in the root.

**Comment by dcolligan** [comment]
@jeromekelleher hm, I hadn't considered that.

In that case, would it make sense to make it a part of the ga4gh package somewhere?  Hm, but even if we did that, we would still need some launching glue in the root, a la server_dev or client_dev.

Ok, I take it back... leaving it in the root is fine.

**Comment by dcolligan** [comment]
(Python does have a relative import syntax, but it only can be used within a package, not across them https://docs.python.org/2/tutorial/modules.html#intra-package-references )

**Comment by Naburimannu** [comment]
Does anybody recognize this travis failure?

```
Searching for avro
Reading https://pypi.python.org/simple/avro/
No local packages or download links found for avro
error: Could not find suitable distribution for Requirement.parse('avro')
The command "python setup.py install" failed and exited with 1 during .
```

I can't just retry Travis jobs, I think because I don't have the Contributor bit.

**Comment by jeromekelleher** [comment]
Nope, haven't seen that one before. PyPI seems to be a weak link in general here, I've seen it refusing connections from Travis a few times.

**Comment by jeromekelleher** [comment]
I reran it there, and it works fine now. We need some bdist_wheels for our requirements we can download from somewhere reliable, this is going to get very tedious otherwise.

**Comment by jeromekelleher** [comment]
OK, I think this is ready to merge. @dcolligan, do you agree?

**Comment by dcolligan** [comment]
@Naburimannu , please squash your commits down to one commit and rebase on top of the current master before resubmitting the branch

After that, I'll merge it.  (or someone else can... here's my +1 after that: +1)

**Comment by Naburimannu** [comment]
Rebased & squashed.
I notice that I was overly-literal when you suggested removing args, and didn't also remove the argument parser, but I'd like to go ahead and land this soonish; the next CL I'm planning in this area will either use or remove it.
